### PR TITLE
Nullable TagTransformers

### PR DESCRIPTION
**Pull request by Kielek**
Follow up to #4282

## Changes

Enable nullable for TagTransformers and its all implementation.
It is used by OTLP project and it is last preparation step to make it nullable.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* ~~[ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes~~
* ~~[ ] Changes in public API reviewed (if applicable)~~

### CODE REVIEW

**Review by pellared — APPROVED** [review]

**Comment by Kielek** [comment]
@TimothyMothra, could you please review?

**Review by TimothyMothra — APPROVED** [review]
LGTM!

File: src/OpenTelemetry/Internal/TagAndValueTransformer.cs [hunk 1]
```diff
@@ -44,7 +47,7 @@ protected TValue TransformValue(object value)
                 return this.TransformFloatingPointValue(Convert.ToDouble(value));
             default:
                 // This could throw an exception. The caller is expected to handle.
-                return this.TransformStringValue(value.ToString());
+                return this.TransformStringValue(value?.ToString());
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Internal/TagAndValueTransformer.cs:50]: There is the `Debug.Assert` on line 28 so it seems like the intention here `null` is not allowed and enforced elsewhere. What I would suggest in that case is not allow a `null` downstream. Just make it:

```diff
- return this.TransformStringValue(value?.ToString());
+ return this.TransformStringValue(value!.ToString()); // The ! is needed for older runtimes where Debug.Assert does not have null decoration. Like netstandard2.0 and net462.

- protected abstract TValue TransformStringValue(string? value);
+ protected abstract TValue TransformStringValue(string value);
```

Comment (Kielek) [reply src/OpenTelemetry/Internal/TagAndValueTransformer.cs:50]: I thing that we have two topics here.
1. `value` detected as null. - I agree that it should not happen. Can we just remove Debug.Assert? The compiler should keep code safe.
2. Changing TransformStringValue contract. Technically. Object.ToString() is defined as `public virtual string? ToString()` (in .NET 6). What is your recommendation for handling such case?

File: src/OpenTelemetry/Internal/TagTransformer.cs [hunk 2]
```diff
@@ -74,7 +81,15 @@ public bool TryTransformTag(KeyValuePair<string, object> tag, out T result, int?
             default:
                 try
                 {
-                    result = this.TransformStringTag(tag.Key, TruncateString(tag.Value.ToString(), maxLength));
+                    var value = tag.Value.ToString();
```
[the remark this answers is not in the document]
Comment (Kielek) [reply src/OpenTelemetry/Internal/TagTransformer.cs]: Fixed: 34484a447a67c216c1d55fdf1e098ef3b0eedb1c

File: src/OpenTelemetry/Internal/TagTransformer.cs [hunk 3]
```diff
@@ -28,9 +33,11 @@ public bool TryTransformTag(KeyValuePair<string, object> tag, out T result, int?
 
         switch (tag.Value)
         {
-            case char:
-            case string:
-                result = this.TransformStringTag(tag.Key, TruncateString(Convert.ToString(tag.Value), maxLength));
+            case char c:
+                result = this.TransformStringTag(tag.Key, TruncateString(c.ToString(), maxLength));
```
[the remark this answers is not in the document]
Comment (Kielek) [reply src/OpenTelemetry/Internal/TagTransformer.cs:38]: Just after merge, I can create separate issue to track this.

**Pull request closed without merging**